File: data-core/src/main/java/org/zzt/note/data/core/entity/NoteNode.java
package org.zzt.note.data.core.entity;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Index;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import org.athena.framework.data.jpa.domain.LogicalDeleteEntity;

/**
 *
 * @author zhouzhitong
 * @since 2026/3/26
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(
        name = "note_node",
        indexes = {
                @Index(name = "idx_note_node_parent_id", columnList = "parent_id"),
                @Index(name = "idx_note_node_note_type", columnList = "note_type"),
                @Index(name = "idx_note_node_parent_sort", columnList = "parent_id,sort")
        }
)
@ToString(callSuper = true, exclude = "meta")
public class NoteNode extends LogicalDeleteEntity {

    /**
     * 父节点ID（根节点为 null）
     */
    @Column(name = "parent_id")
    private Long parentId;

    /**
     * 标题
     */
    @Column(nullable = false, length = 200)
    private String title;

    /**
     * 笔记模板类型（核心字段）
     */
    @Column(name = "note_type", length = 50)
    private String noteType;

    /**
     * 排序
     */
    @Column(nullable = false)
    private Integer sort = 0;

    /**
     * 内容（JSON）
     */
    @Column(columnDefinition = "TEXT")
    private String content;

    /**
     * 节点附加信息（subject/icon/tags）
     */
    @OneToOne(mappedBy = "node", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private NoteNodeMeta meta;

    public void setMeta(NoteNodeMeta meta) {
        if (this.meta != null && this.meta != meta) {
            this.meta.setNode(null);
        }
        this.meta = meta;
        if (meta != null && meta.getNode() != this) {
            meta.setNode(this);
        }
    }

}

File: note-ui/src/model/note/noteNode.js
/**
 * @typedef {Object} NoteTagDTO
 * @property {number|null} id
 * @property {string} name
 */

/**
 * @typedef {Object} NoteNodeMetaDTO
 * @property {Object|null} node
 * @property {string} icon
 * @property {NoteTagDTO[]} tags
 * @property {string} subject
 */

/**
 * @typedef {Object} NoteNodeDTO
 * @property {number|null} id
 * @property {number|null} parentId
 * @property {string} title
 * @property {string} noteType
 * @property {number} sort
 * @property {string} content
 * @property {NoteNodeMetaDTO} meta
 */

/**
 * 与后端 NoteNodeDTO 对齐的默认结构。
 * @returns {NoteNodeDTO}
 */
export function createDefaultNoteNode() {
  return {
    id: null,
    parentId: null,
    title: "",
    noteType: "",
    sort: 0,
    content: "",
    meta: {
      node: null,
      icon: "",
      tags: [],
      subject: "",
    },
  };
}

/**
 * 页面展示用的示例数据（结构与 NoteNodeDTO 对齐）。
 * @returns {NoteNodeDTO}
 */
export function createMockNoteNode() {
  return {
    id: 20260327001,
    parentId: null,
    title: "これは私の本です",
    noteType: "VOCAB_NOTE",
    sort: 10,
    content: JSON.stringify({
      paragraphs: [
        "这是一个 noteNode 的展示区域，用来承载当前节点的正文内容。",
        "你可以在这里展示文本、说明、例句、结构化内容，或者后续扩展为富文本渲染区域。",
        "如果你的 noteNode 是“特定业务模板笔记”，这里还可以根据 noteType 动态切换不同内容区域。例如：单词卡片、知识点、题目、复习记录、案例说明等。",
      ],
      bullets: [
        "支持展示节点标题",
        "支持展示类型、学科、排序等基础属性",
        "支持后续扩展标签、状态、模板字段",
      ],
    }),
    meta: {
      node: null,
      icon: "📘",
      tags: [
        { id: 1, name: "N5" },
        { id: 2, name: "基础句型" },
        { id: 3, name: "指示代词" },
      ],
      subject: "日语",
    },
  };
}
